File: java/src/main/java/vsdk/toolkit/io/context/GeometryAssemblyContext.java
package vsdk.toolkit.io.context;

import java.util.ArrayList;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.skin.Geometry;
import vsdk.toolkit.skin.Patch;
import vsdk.toolkit.skin.Vertex;

public class GeometryAssemblyContext {
    public static final int MAXIMUM_GEOMETRY_STACK_DEPTH = 100;

    public String currentVertexName;
    public int geometryStackHeadIndex;
    public ArrayList<Geometry>[] geometryStack;

    // Those lists are transferred to MeshSurface objects and should not be deleted on transfer.
    public ArrayList<Vector3D> currentPointList;
    public ArrayList<Vector3D> currentNormalList;
    public ArrayList<Vertex> currentVertexList;
    public ArrayList<Patch> currentFaceList;
    public ArrayList<Geometry> currentGeometryList;
    public String currentObjectName;
    public boolean inSurface;
    public boolean inComplex;
    public boolean warpConeEnds;
    public ArrayList<Geometry> allGeometries;

    // Return model geometry output
    public ArrayList<Geometry> geometries;

    @SuppressWarnings("unchecked")
    public GeometryAssemblyContext() {
        currentVertexName = null;
        geometryStackHeadIndex = 0;
        geometryStack = (ArrayList<Geometry>[])new ArrayList[MAXIMUM_GEOMETRY_STACK_DEPTH];
        currentPointList = null;
        currentNormalList = null;
        currentVertexList = null;
        currentFaceList = null;
        currentGeometryList = null;
        currentObjectName = null;
        inSurface = false;
        inComplex = false;
        warpConeEnds = false;
        allGeometries = new ArrayList<>();
        geometries = null;
    }

    public void destroy() {
        currentObjectName = null;
        if (allGeometries != null) {
            allGeometries.clear();
            allGeometries = null;
        }
    }
}

File: typescript/src/main/vsdk/toolkit/io/context/GeometryAssemblyContext.ts
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { Geometry } from "../../skin/Geometry";
import { Patch } from "../../skin/Patch";
import { Vertex } from "../../skin/Vertex";

export class GeometryAssemblyContext {
  public static readonly MAXIMUM_GEOMETRY_STACK_DEPTH = 100;

  public currentVertexName: string | null;
  public geometryStackHeadIndex: number;
  public geometryStack: Array<Geometry[] | null>;

  public currentPointList: Vector3D[] | null;
  public currentNormalList: Vector3D[] | null;
  public currentVertexList: Vertex[] | null;
  public currentFaceList: Patch[] | null;
  public currentGeometryList: Geometry[] | null;
  public currentObjectName: string | null;
  public inSurface: boolean;
  public inComplex: boolean;
  public warpConeEnds: boolean;
  public allGeometries: Geometry[] | null;

  public geometries: Geometry[] | null;

  public constructor() {
    this.currentVertexName = null;
    this.geometryStackHeadIndex = 0;
    this.geometryStack = new Array<Geometry[] | null>(GeometryAssemblyContext.MAXIMUM_GEOMETRY_STACK_DEPTH).fill(null);
    this.currentPointList = null;
    this.currentNormalList = null;
    this.currentVertexList = null;
    this.currentFaceList = null;
    this.currentGeometryList = null;
    this.currentObjectName = null;
    this.inSurface = false;
    this.inComplex = false;
    this.warpConeEnds = false;
    this.allGeometries = [];
    this.geometries = null;
  }

  public destroy(): void {
    this.currentObjectName = null;
    if (this.allGeometries !== null) {
      this.allGeometries.length = 0;
      this.allGeometries = null;
    }
  }
}
